### PR TITLE
fix(pam/gdm) model/module improvements

### PR DESCRIPTION
**Pull request by 3v1n0**
Various improvements to the module when integrating with gdm in the wild

### CODE REVIEW

**Comment by 3v1n0** [comment]
I reverted the change on using pointer receiver, things still go fine and I added some more testing.

I think the most reliable way to handle stopping operations is using msg filtering (also as per bubbletea docs), so I've used that. I personally would have preferred to expose only a `QuitHandler` function instead that the whole `MsgFilter`, but for the way things are, if we abstract that too much then we need to duplicate code and test things less.

However, I'm wondering if even `endSession` should be handled so that it happens every time a `tea.QuitMsg` happens

File: debian/authd.gdm-authd.pam
Comment (denisonbarbosa) [inline debian/authd.gdm-authd.pam:3]: I'm not a big fan of using `default=die` in our module config as we could lock out the system entirely if `authd` can't be executed. Can we think about something better? 

cc. @didrocks

File: pam/internal/adapter/gdmmodel.go
Comment (denisonbarbosa) [inline pam/internal/adapter/gdmmodel.go:26]: This deserves a comment explaining why we need this

File: pam/pam.go
Comment (denisonbarbosa) [inline pam/pam.go:95]: No need for this whitespace. The check below evaluates the returned error from h.handleAuthRequest, so it should be scoped with the function call.

File: pam/pam.go
Comment (denisonbarbosa) [inline pam/pam.go:101]: Remember about the happy path...

```Go
if err != nil && !errors.Is(err, pam.ErrIgnore) {
    return err        
}

return mTx.SetData(authenticatedKey, struct{}{})
```

File: pam/pam.go
Comment (denisonbarbosa) [inline pam/pam.go:116]: If you are not using this anywhere else, you should just parse it right with the `newClient` call

File: pam/pam.go
Comment (denisonbarbosa) [inline pam/pam.go:144]:
```suggestion
	client, closeConn, err := newClient(parseArgs(args))
```

**Review by denisonbarbosa — CHANGES_REQUESTED** [review]
A few suggestions, but looking good!

File: debian/authd.gdm-authd.pam
Comment (didrocks) [reply debian/authd.gdm-authd.pam:3]: Anyway, we are including common-auth just below, to authd will be already provided in common-auth, no? Also, the fact that we have `success=2`, hardcoded, where actually the amount of steps we want to jump over depends on the one in common-auth (if the content is only included in that one) seems fragile to me?

Comment (3v1n0) [reply debian/authd.gdm-authd.pam:3]: > I'm not a big fan of using `default=die` in our module config as we could lock out the system entirely if `authd` can't be executed. Can we think about something better?

Well, if we fail, really we should be locked down when using gdm with authd mode. In case it fallback to local broker then we're ignoring, so that's handled.

In any case, in GDM you can technically always switch pam profile to use, so for example smart-cards or fingerprint profiles will always be running despite this failing.

> Anyway, we are including common-auth just below, to authd will be already provided in common-auth, no? Also, the fact that we have success=2, hardcoded, where actually the amount of steps we want to jump over depends on the one in common-auth (if the content is only included in that one) seems fragile to me?

Oh, yeah, in fact it seems I eventually pushed the old version of this file -_-, we've in fact to use `substack` to handle this in this way.

This file is basically based on `gdm-smartcard-sssd-exclusive` pam config that I wrote some time ago and it's quite well tested by autopkgtests too.

An alternative to avoid using jumps (but repetitions) is instead to do something like this:

```
#%PAM-1.0
auth    [success=ok user_unknown=ignore default=bad] pam_succeed_if.so user != root quiet_success
auth    [success=ok ignore=3 default=die] pam_authd.so
auth    requisite       pam_nologin.so
auth    optional        pam_gnome_keyring.so
auth    sufficient      pam_permit.so

# If authd is ignoring the request we fallback to default auth
@include common-auth
auth    requisite       pam_nologin.so
auth    optional        pam_gnome_keyring.so
```

Comment (didrocks) [reply debian/authd.gdm-authd.pam:3]: I think this is more reasonable if we don’t include common-auth and do the same than other gdm specific processes.

Comment (3v1n0) [reply debian/authd.gdm-authd.pam]: > I think this is more reasonable if we don’t include common-auth and do the same than other gdm specific processes, so +1 on the correct file (happy to have spotted it!)

No check the edit, it was fine already :-D. For some reason I thought the pushed file was using `@include` but it didn't, it was using `substack` and so was fine.

And IMHO it's better to use it compared to including.

Comment (3v1n0) [reply debian/authd.gdm-authd.pam]: Mostly because we'd be more conformant to what we already do for other gdm stacks, and being consistent with the others, makes it easier to debug or address potential issues.

Comment (didrocks) [reply debian/authd.gdm-authd.pam]: Ah, so the fix was to change the `success=2` to 1 :)

File: pam/pam.go
Comment (3v1n0) [reply pam/pam.go:116]: Not here, but this is something that is going to be used at later point, (see for example #233), so I preferred to lave it as a variable not to have to refactor it again at later point.

File: debian/authd.gdm-authd.pam
Comment (didrocks) [reply debian/authd.gdm-authd.pam]: agreed on following the rest of the gdm stacks.

Comment (3v1n0) [reply debian/authd.gdm-authd.pam]: > Ah, so the fix was to change the `success=2` to 1 :)

No, it was already working as this, but I noticed that _being consistent_ I had copied too much from gdm one :-D, were we do an extra step and so we've to skip 2 times.

By having `2` here we'd missed using `pam_nologin.so` which isn't super-needed but still something we should do

Comment (didrocks) [reply debian/authd.gdm-authd.pam]: got it!

**Review by denisonbarbosa — APPROVED** [review]
LGTM!

File: pam/pam.go
Comment (denisonbarbosa) [reply pam/pam.go]: Yeah, but that would also happen on your previous logic so it's not the happy path's fault 😉

Comment (3v1n0) [reply pam/pam.go]: Nope, it wasn't (in fact CI wasn't fail as per https://github.com/ubuntu/authd/actions/runs/8069328331/job/22044497208?pr=226)

Since in case of ignore error it was returning the one on outer scope. However, I'm happier with the happy path too :)

Comment (denisonbarbosa) [reply pam/pam.go]: Oh, interesting... So if we `if err := ...` it already scopes the value and doesn't shadow the previous variable. Good to know.

File: pam/internal/adapter/gdmmodel.go
Comment (didrocks) [inline pam/internal/adapter/gdmmodel.go:296]: isn’t that a busy loop? If so, I think we should have a timeout with channel.

File: pam/internal/adapter/gdmmodel.go
Comment (didrocks) [inline pam/internal/adapter/gdmmodel.go:295]: prefer time.After() then.

File: pam/internal/adapter/gdmmodel_test.go
Comment (didrocks) [inline pam/internal/adapter/gdmmodel_test.go:2000]: nitpick: one character should be enough in the context, so `r`.

File: pam/pam.go
Comment (didrocks) [inline pam/pam.go:39]: The key is rather `already-authenticated-flag` or something along the line to explain this is to prevent reentry?

File: pam/pam.go
Comment (didrocks) [inline pam/pam.go:89]: I would do stricter arg parsing in `parseArgs`, at least, warning on unknown keys. That would at least document in one place our supported keys.

File: pam/pam.go
Comment (didrocks) [inline pam/pam.go:268]: that one for instance!

**Review by didrocks — CHANGES_REQUESTED** [review]
Some minor comments, but I think that would be great to have them fixed before merging!

File: pam/internal/adapter/gdmmodel.go
Comment (3v1n0) [reply pam/internal/adapter/gdmmodel.go:296]: Yeah, I was thinking about it... It's also true that if we fail here we should probably panic, no?

Comment (didrocks) [reply pam/internal/adapter/gdmmodel.go:296]: We should never panic is a library, always return an error. Don’t panic in Go, in general

File: pam/internal/adapter/gdmmodel.go
Comment (3v1n0) [reply pam/internal/adapter/gdmmodel.go:295]: It's twp different phases though, first I'd give gdm some time to complete the operations, then we force waiting. So first wait should still be with time.after()? sleep doesn't seem blocking other goroutines (as others are managed by bubbletea).

File: pam/pam.go
Comment (3v1n0) [reply pam/pam.go:89]: Ok.... This is done now, but for testing it I had to steal few commits from #233

File: pam/internal/adapter/gdmmodel.go
Comment (3v1n0) [reply pam/internal/adapter/gdmmodel.go]: Well, as programmer error... Even though this is not the case, so I'll warn only.

File: pam/internal/adapter/gdmmodel.go
Comment (3v1n0) [reply pam/internal/adapter/gdmmodel.go]: Eh, also because using time after.... `S1037: should use time.Sleep instead of elaborate way of sleeping (gosimple)` (which was my guess too :))

File: pam/pam.go
Comment (denisonbarbosa) [reply pam/pam.go:102]: You didn't need to assert on the logging messages. IMHO, the test for it should only be in the sense of "Unsupported argument gets ignored" and then proceed with a simple authentication (rather than stopping in the middle of nowhere as the current tape file does).

Comment (3v1n0) [reply pam/pam.go:102]: Ok, done... I feel it's incomplete testing, but I also agree checking on logs isn't the best.

At least, it would have been better with a proper log handler and ensure that we call it, but that's fine for now.

File: pam/integration-tests/integration_test.go
Comment (denisonbarbosa) [inline pam/integration-tests/integration_test.go:50]: Forgot to remove the additional field (I do this everytime as well :smiling_face_with_tear: )

File: pam/integration-tests/integration_test.go
Comment (denisonbarbosa) [inline pam/integration-tests/integration_test.go:274]: I think we don't care about this now that we don't assert on the logs anymore, right?

File: pam/integration-tests/testdata/TestCLIAuthenticate/golden/authenticate_with_warnings_on_unsupported_arguments
Comment (denisonbarbosa) [inline pam/integration-tests/testdata/TestCLIAuthenticate/golden/authenticate_with_warnings_on_unsupported_arguments:170]: That's exactly what I talked about. Thanks!

File: pam/integration-tests/testdata/TestCLIAuthenticate/golden/warns_on_unsupported_arguments
Comment (denisonbarbosa) [inline pam/integration-tests/testdata/TestCLIAuthenticate/golden/warns_on_unsupported_arguments:1]: Leftover file from the previous test

**Review by denisonbarbosa — APPROVED** [review]
A couple of leftovers from the previous changes, but easily addressable. Well done!

File: pam/integration-tests/testdata/TestCLIAuthenticate/golden/warns_on_unsupported_arguments
Comment (3v1n0) [reply pam/integration-tests/testdata/TestCLIAuthenticate/golden/warns_on_unsupported_arguments]: Oh, damn... Yeah, it wouldn't have happened if I'd had rebased as I would have dropped the commit at all, but thanks for spotting it!

File: pam/integration-tests/integration_test.go
Comment (3v1n0) [reply pam/integration-tests/integration_test.go:272]: Well this was part of #233 and while we don't care per se it's quite useful in case of failures since it gives us the cli logs to inspect.

So maybe still worth to have it in case we have unexpected failures in launchpad builds?

Comment (denisonbarbosa) [reply pam/integration-tests/integration_test.go:272]: You are saving the logs to a temporary dir that gets removed when the test ends, so we don't get anything at all. We could add this logic to the SaveArtifactsForDebug function that saves the artefacts on a temporary dir that is not related to the tests themselves, but it will require more changes and it's not a priority right now. So I think removing this the best option for now.

Comment (3v1n0) [reply pam/integration-tests/integration_test.go:272]: Indeed the artifact was added to the artifact dir, plus as you can see in the cleanup func we do `t.Log(string(out))`, so it's logged (try this branch with `-v`) and on failure is shown.

So changes are all there already (unless I forgot to pick something), but still let me know if you prefer to keep this for later.

Comment (3v1n0) [reply pam/integration-tests/integration_test.go:272]: Indeed as per https://github.com/ubuntu/authd/pull/226/commits/dd51fdfbec77e0f932c313918435a9906d0aa644 it should do it all

File: pam/internal/adapter/gdmmodel.go
Comment (didrocks) [inline pam/internal/adapter/gdmmodel.go:300]: This is still a busy loop, no? it will spin instead of waiting for a number of mutex release?

File: pam/integration-tests/integration_test.go
Comment (didrocks) [reply pam/integration-tests/integration_test.go:272]: I think from the disussion this should be separated, no?

**Review by didrocks — CHANGES_REQUESTED** [review]
We are not extracting the socket options, do we?

Other than one, my biggest remaining comment is on the busy loop itself, or am I missing anything?

File: pam/internal/adapter/gdmmodel.go
Comment (3v1n0) [reply pam/internal/adapter/gdmmodel.go:300]: Oh well... Indeed it is, but I think we need to wait here, and per se a gdm conversation can be initiated by multiple actors.

I can refactor this though, ideally I wanted to keep as simple as possible since this happens when we're exiting only and we don't have really resources constraints.

Comment (3v1n0) [reply pam/internal/adapter/gdmmodel.go:300]: So something like...

```diff
diff --git a/pam/internal/adapter/gdmmodel.go b/pam/internal/adapter/gdmmodel.go
index 451612f..20fda2f 100644
--- a/pam/internal/adapter/gdmmodel.go
+++ b/pam/internal/adapter/gdmmodel.go
@@ -297,8 +297,7 @@ func (m gdmModel) stopConversations() gdmModel {
 
 	wait := make(chan struct{})
 	go func() {
-		for gdm.ConversationInProgress() {
-		}
+		gdm.WaitForConversationsStopped()
 		wait <- struct{}{}
 	}()
 
diff --git a/pam/internal/gdm/conversation.go b/pam/internal/gdm/conversation.go
index a32bc5c..e2f8172 100644
--- a/pam/internal/gdm/conversation.go
+++ b/pam/internal/gdm/conversation.go
@@ -6,27 +6,49 @@ import (
 	"context"
 	"errors"
 	"fmt"
-	"sync/atomic"
+	"sync"
 
 	"github.com/msteinert/pam/v2"
 	"github.com/ubuntu/authd/internal/log"
 )
 
-var conversations atomic.Int32
+var conversations int
+var conversationsMu = sync.Mutex{}
+var conversationsCond = sync.Cond{L: &conversationsMu}
 
 // ConversationInProgress checks if conversations are currently active.
 func ConversationInProgress() bool {
-	return conversations.Load() > 0
+	return false
+	// return conversations.Load() > 0
+}
+
+func WaitForConversationsStopped() {
+	conversationsMu.Lock()
+	defer conversationsMu.Unlock()
+
+	for conversations != 0 {
+		conversationsCond.Wait()
+	}
 }
 
 func sendToGdm(pamMTx pam.ModuleTransaction, data []byte) ([]byte, error) {
-	conversations.Add(1)
-	defer conversations.Add(-1)
 	binReq, err := NewBinaryJSONProtoRequest(data)
 	if err != nil {
 		return nil, err
 	}
 	defer binReq.Release()
+
+	conversationsMu.Lock()
+	conversations++
+	conversationsMu.Unlock()
+
+	defer func() {
+		conversationsMu.Lock()
+		conversations--
+		conversationsCond.Broadcast()
+		conversationsMu.Unlock()
+	}()
+
 	res, err := pamMTx.StartConv(binReq)
 	if err != nil {
 		return nil, err
```

It seems to me that it adds lot of complexity to an operation that happens a lot during runtime (e.g. all the polling...) while spin-locking it's an one time thing only during shutdown...

Comment (3v1n0) [reply pam/internal/adapter/gdmmodel.go]: Ok, pushed the version we agree'd in pvt.

**Review by didrocks — APPROVED** [review]
Thanks for the fixes!

**Pull request merged**
